File: src-lang/dart/net/http_fetch.clj
(ns dart.net.http-fetch
  (:require [hara.lang :as l]
            [xt.lang.common-protocol :refer [defimpl.xt]]))

(l/script :dart
  {:require [[xt.lang.spec-base :as xt]
             [xt.lang.spec-promise :as promise]
             [xt.lang.common-protocol :as protocol]
             [xt.net.http-fetch :as fetch]
             [xt.net.http-util :as util]]
   :import [["dart:convert" :as convert]
            ["dart:io" :as io]]})

(defn.dt request-http-raw
  [client input]
  (var #{raw} client)
  (:= raw (or raw {}))
  (var request (fetch/prepare-input client input))
  (var request-fn (xt/x:get-key raw "request"))
  (if (xt/x:is-function? request-fn)
    (do (var output (request-fn request {}))
        (return (util/response-normalize output)))
    (do (var http-client (io.HttpClient))
        (var uri (. Uri (parse (xt/x:get-key request "url"))))
        (return
         (. (. http-client (openUrl (xt/x:get-key request "method") uri))
            (then
             (fn [req]
              (xt/for:object [[k v] (xt/x:get-key request "headers")]
                (. (. req headers) (set k v)))
              (when (xt/x:not-nil? (xt/x:get-key request "body"))
                (. req (write (xt/x:get-key request "body"))))
              (var response-future (. req (close)))
              (return
               (. response-future
                  (then
                   (fn [res]
                     (var body-future
                          (. (. res (transform (. (. convert utf8) decoder)))
                             (join)))
                     (return
                      (. body-future
                         (then
                          (fn [body]
                            (return {"status" (. res statusCode)
                                     "headers" {}
                                     "body" (util/decode-body body)}))))))))))))))))

(defn.dt request-http
  [client input]
  (var handler (fetch/prepare-middleware client -/request-http-raw))
  (return (handler client input)))

(defimpl.xt ^{:lang :dart}
  DartHttpFetchClient
  [defaults middleware]
  fetch/IHttpClient
  {fetch/request-http -/request-http})

(defn.dt create
  [defaults middleware]
  (return
   (-/DartHttpFetchClient (or defaults {}) (or middleware
                                               [fetch/wrap-prepare-input]))))

File: packages-gen/dart/libs/xt-net/lib/net/http-fetch.dart
import 'dart:io' as io;
import 'dart:convert' as convert;
import 'package:xtalk_lang/common-protocol.dart' as protocol;
import 'package:xtalk_net/http-fetch.dart' as fetch;
import 'package:xtalk_net/http-util.dart' as util;
import 'dart:io';






DartHttpFetchClient(defaults, middleware) {
  (__globals__["xt.lang.common_protocol/IMPLEMENTATIONS"] ??= <dynamic, dynamic>{})["dart.net.http_fetch/DartHttpFetchClient"] = true;
  protocol.register_protocol_impl(
    fetch.IHttpClient["on"],
    "dart.net.http_fetch/DartHttpFetchClient",
    <dynamic, dynamic>{"request_http":request_http}
  );
  return <dynamic, dynamic>{
    "::":"dart.net.http_fetch/DartHttpFetchClient",
    "::/protocols":<dynamic>[fetch.IHttpClient["on"]],
    "::/protocol-impls":<dynamic, dynamic>{
        fetch.IHttpClient["on"]:<dynamic, dynamic>{"request_http":request_http}
      },
    "defaults":defaults,
    "middleware":middleware
  };
}

request_http_raw(client, input) {
  var raw = client["raw"];
  raw = (raw ?? <dynamic, dynamic>{});
  var request = fetch.prepare_input(client,input);
  var request_fn = raw["request"];
  if((request_fn.runtimeType).toString().contains("Function") || (request_fn.runtimeType).toString().contains("=>") || (request_fn).toString().startsWith("Closure")){
    var output = Function.apply(
      (request_fn as Function),
      <dynamic>[request,<dynamic, dynamic>{}]
    );
    return util.response_normalize(output);
  }
  else{
    var http_client = io.HttpClient();
    var uri = Uri.parse(request["url"]);
    var output_future = http_client.openUrl(request["method"],uri).then((req) {
      for(var entry_51592 in request["headers"].entries){
        var k = entry_51592.key;
        var v = entry_51592.value;
        req.headers.set(k,v);
      };
      if(null != request["body"]){
        var body_bytes = convert.utf8.encode(request["body"]);
        req.contentLength = body_bytes.length;
        req.add(body_bytes);
      }
      var response_future = req.close();
      return response_future.then((res) {
        var body_future = res.transform(convert.utf8.decoder).join();
        return body_future.then((body) {
          return <dynamic, dynamic>{
            "status":res.statusCode,
            "headers":<dynamic, dynamic>{},
            "body":util.decode_body(body)
          };
        });
      });
    });
    return output_future.whenComplete(() {
      return http_client.close();
    });
  }
}

request_http(client, input) {
  var handler = fetch.prepare_middleware(client,request_http_raw);
  return Function.apply((handler as Function),<dynamic>[client,input]);
}

create(defaults, middleware) {
  return DartHttpFetchClient(
    defaults ?? <dynamic, dynamic>{},
    middleware ?? <dynamic>[fetch.wrap_prepare_input]
  );
}